### DISCUSSION TITLE
Should P/R/G update URLs in forms as well?

### DISCUSSION THREAD

**Discussion by zraly** [jiripudil/Naja#305] [Support] [answered]
When using this:
```
$this->payload->postGet = TRUE;
$this->payload->url = $this->link('this');
```

Should also `action` form parameters be updated? Nette generates form actions automatically, and these should be IMHO updated. But there can be form with custom different action. How to solve this?

**Comment by jiripudil** [comment, marked as answer]
Hello, what's your particular use case? Do your forms point somewhere they shouldn't, and can they not be processed as a result? If their action needs to change, I'd argue they should be updated in a snippet 🤔

**Reply by zraly** [reply]
My use case: 
- Let's say I have a blog admin. 
- I go to a particular post to edit it.
- The URL changes to `/admin/post/old-post-slug`. 
- There is a form for editing the post with action set to `/admin/post/old-post-slug` of course (Nette generates that).
- I edit the form – changing the post's slug to `new-post-slug`.
- Because the URL has changed, after form process I send 
```
$this->payload->postGet = TRUE;
$this->payload->url = '/admin/post/new-post-slug';
```
- But the form still have action `/admin/post/old-post-slug`, so any other form submitting ends with error, because the URL does not exists.

**Reply by jiripudil** [reply]
Thanks for the clarification. I think the best solution is indeed wrapping the forms in a snippet and redrawing them with the updated action URL. Or, If you want to do it client-side, you could use an extension, something along these lines could work:

```js
class UpdateActionExtension {
    initialize(naja) {
        naja.addEventListener('start', (event) => {
            event.detail.options.lastUrl = location.href;
        });

        naja.addEventListener('success', (event) => {
            if (event.detail.payload.postGet && event.detail.payload.url) {
                for (const form of document.querySelectorAll('form')) {
                    if (new URL(form.action, location.href).href === event.detail.options.lastUrl) {
                        form.action = event.detail.payload.url;
                    }
                }
            }
        });
    }
}
```

I don't think this is something Naja should do automatically because, as you've mentioned, it would be quite difficult to account for all use cases. I can imagine it could be pretty confusing to developers.

**Reply by zraly** [reply]
I agree. I already did it the way you offered. Thanks.
